### PR TITLE
Gc optimize

### PR DESCRIPTION
**Pull request by Ashuaidehao**
Reduce unnecessary GC.

### CODE REVIEW

File: src/neo-vm/ExecutionEngine.cs [hunk 1]
```diff
@@ -1577,7 +1578,11 @@ public StackItem Pop()
         /// </summary>
         protected virtual void PostExecuteInstruction()
         {
-            if (ReferenceCounter.CheckZeroReferred() > Limits.MaxStackSize)
+            if (ReferenceCounter.Count > Limits.MaxStackSize)
+            {
+                ReferenceCounter.CheckZeroReferred();
+            }
+            if (ReferenceCounter.Count > Limits.MaxStackSize)
                 throw new InvalidOperationException($"MaxStackSize exceed: {ReferenceCounter.Count}");
```
Comment (shargon) [inline src/neo-vm/ExecutionEngine.cs:1586]:
```suggestion
            if (ReferenceCounter.Count > Limits.MaxStackSize && ReferenceCounter.CheckZeroReferred() > Limits.MaxStackSize)
                throw new InvalidOperationException($"MaxStackSize exceed: {ReferenceCounter.Count}");
```

File: src/neo-vm/ExecutionEngine.cs [hunk 2]
```diff
@@ -129,6 +129,7 @@ public virtual VMState Execute()
                 State = VMState.NONE;
             while (State != VMState.HALT && State != VMState.FAULT)
                 ExecuteNext();
+            ReferenceCounter.CheckZeroReferred();
```
Comment (shargon) [inline src/neo-vm/ExecutionEngine.cs:132]: Why this line?

Comment (Ashuaidehao) [reply src/neo-vm/ExecutionEngine.cs:132]: To make sure GC occur at least once.

**Comment by vang1ong7ang** [comment]
@Ashuaidehao  are you serious? :new_moon_with_face: :new_moon_with_face: :new_moon_with_face:

**Comment by vang1ong7ang** [comment]
have you heared a story named [Boiling Frog](https://en.wikipedia.org/wiki/Boiling_frog) :new_moon_with_face: :new_moon_with_face:

**Comment by dusmart** [comment]
> have you heared a story named [Boiling Frog](https://en.wikipedia.org/wiki/Boiling_frog) 🌚 🌚

1. add A to array B
2. remove A from B
3. repeat 1~2 thousands of times
4. constructt an O(n**2)  `CheckZeroReferred` such as <https://github.com/neo-project/neo/issues/2532>

**Comment by vang1ong7ang** [comment]
@dusmart yours is not optimal :new_moon_with_face: better to make extreme use of the `O(n^2)` feature

**Comment by erikzhang** [comment]
What problem does this PR solve?

**Comment by Ashuaidehao** [comment]
Reduce the frequency of GC may get performance improvement as a whole. If GC only used for check max item limit, we donot need GC after every instruction.

**Comment by vang1ong7ang** [comment]
@Ashuaidehao it doesn't make sense, man, unless GC is less than `O(n)` like `O(1)`.

**Comment by dusmart** [comment]
> Reduce unnecessary GC.

In my opinion, every GC is necessary.

> 今日事今日毕。
> 明日复明日，明日何其多。

**Comment by dusmart** [comment]
> Reduce the frequency of GC may get performance improvement as a whole. If GC only used for check max item limit, we donot need GC after every instruction.

GC is used for GC.

**Comment by vang1ong7ang** [comment]
under current code base, this PR destroys the GC imo.

**Pull request closed without merging**